src/validation/rules/UniqueArgumentDefinitionNamesRule.ts

```diff
@@ -0,0 +1,91 @@
+import { GraphQLError } from '../../error/GraphQLError';
+
+import type { ASTVisitor } from '../../language/visitor';
+import type {
+  NameNode,
+  FieldDefinitionNode,
+  InputValueDefinitionNode,
+} from '../../language/ast';
+
+import type { SDLValidationContext } from '../ValidationContext';
+
+/**
+ * Unique argument definition names
+ *
+ * A GraphQL Object or Interface type is only valid if all its fields have uniquely named arguments.
+ * A GraphQL Directive is only valid if all its arguments are uniquely named.
+ */
+export function UniqueArgumentDefinitionNamesRule(
+  context: SDLValidationContext,
+): ASTVisitor {
+  return {
+    DirectiveDefinition(directiveNode) {
+      // istanbul ignore next (See: 'https://github.com/graphql/graphql-js/issues/2203')
+      const argumentNodes = directiveNode.arguments ?? [];
+
+      return checkArgUniqueness(`@${directiveNode.name.value}`, argumentNodes);
+    },
+    InterfaceTypeDefinition: checkArgUniquenessPerField,
+    InterfaceTypeExtension: checkArgUniquenessPerField,
+    ObjectTypeDefinition: checkArgUniquenessPerField,
+    ObjectTypeExtension: checkArgUniquenessPerField,
+  };
+
+  function checkArgUniquenessPerField(typeNode: {
+    readonly name: NameNode;
+    readonly fields?: ReadonlyArray<FieldDefinitionNode>;
+  }) {
+    const typeName = typeNode.name.value;
+
+    // istanbul ignore next (See: 'https://github.com/graphql/graphql-js/issues/2203')
+    const fieldNodes = typeNode.fields ?? [];
+
+    for (const fieldDef of fieldNodes) {
+      const fieldName = fieldDef.name.value;
+
+      // istanbul ignore next (See: 'https://github.com/graphql/graphql-js/issues/2203')
+      const argumentNodes = fieldDef.arguments ?? [];
+
+      checkArgUniqueness(`${typeName}.${fieldName}`, argumentNodes);
+    }
+
+    return false;
+  }
+
+  function checkArgUniqueness(
+    parentName: string,
+    argumentNodes: ReadonlyArray<InputValueDefinitionNode>,
+  ) {
+    const seenArgs = groupBy(argumentNodes, (arg) => arg.name.value);
+
+    for (const [argName, argNodes] of seenArgs) {
+      if (argNodes.length > 1) {
+        context.reportError(
+          new GraphQLError(
+            `Argument "${parentName}(${argName}:)" can only be defined once.`,
+            argNodes.map((node) => node.name),
+          ),
+        );
+      }
+    }
+
+    return false;
+  }
+}
+
+function groupBy<K, T>(
+  list: ReadonlyArray<T>,
+  keyFn: (item: T) => K,
+): Map<K, Array<T>> {
+  const result = new Map<K, Array<T>>();
+  for (const item of list) {
+    const key = keyFn(item);
+    const group = result.get(key);
+    if (group === undefined) {
```

Comment:
Given `![] === false`, I think this can just be:
```suggestion
    if (!group) {
```